### PR TITLE
WritePrepared Txn: update compaction_iterator_test and db_iterator_test

### PR DESCRIPTION
**Pull request by yiwu-arbug**
Summary:
Update compaction_iterator_test with write-prepared transaction DB related tests. Transaction related tests are group in CompactionIteratorWithSnapshotCheckerTest. The existing test are duplicated to make them also test with dummy SnapshotChecker that will say every key is visible to every snapshot (this is okay, we still compare sequence number to verify visibility). Merge related tests are disabled and will be revisit in another PR.

Existing db_iterator_tests are also duplicated to test with dummy read_callback that will say every key is committed.

Test Plan:
Run the tests.

### CODE REVIEW

File: db/compaction_iterator_test.cc [hunk 1]
```diff
@@ -159,61 +173,129 @@ class FakeCompaction : public CompactionIterator::CompactionProxy {
...
+      CompactionFilter* compaction_filter = nullptr,
+      bool bottommost_level = false,
+      SequenceNumber earliest_write_conflict_snapshot = kMaxSequenceNumber) {
+    InitIterators(input_keys, input_values, {}, {}, kMaxSequenceNumber,
+                  last_committed_seq, merge_operator, compaction_filter,
+                  bottommost_level, earliest_write_conflict_snapshot);
+    c_iter_->SeekToFirst();
+    for (size_t i = 0; i < expected_keys.size(); i++) {
+      std::string info = "i = " + ToString(i);
+      ASSERT_TRUE(c_iter_->Valid()) << info;
```
Comment (maysamyabandeh) [inline db/compaction_iterator_test.cc]: What does "<< std::string" do here?

File: db/compaction_iterator_test.cc [hunk 2]
```diff
@@ -159,61 +173,129 @@ class FakeCompaction : public CompactionIterator::CompactionProxy {
...
+      ASSERT_EQ(expected_values[i], c_iter_->value().ToString()) << info;
+      c_iter_->Next();
+    }
+    ASSERT_FALSE(c_iter_->Valid());
+  }
 
   const Comparator* cmp_;
   const InternalKeyComparator icmp_;
   std::vector<SequenceNumber> snapshots_;
+  std::unordered_map<SequenceNumber, SequenceNumber> snapshot_map_;
```
Comment (maysamyabandeh) [inline db/compaction_iterator_test.cc]: Would be great to add a comment of what is mapped to what.

File: db/compaction_iterator_test.cc [hunk 3]
```diff
@@ -565,6 +649,300 @@ TEST_F(CompactionIteratorTest, SingleMergeOperand) {
...
+TEST_P(CompactionIteratorTest, RemoveDeletionAtBottomLevel) {
+  AddSnapshot(1);
+  RunTest({test::KeyStr("a", 1, kTypeDeletion),
+           test::KeyStr("b", 2, kTypeDeletion)},
+          {"", ""}, {test::KeyStr("b", 2, kTypeDeletion)}, {""},
+          kMaxSequenceNumber /*last_commited_seq*/, nullptr /*merge_operator*/,
+          nullptr /*compaction_filter*/, true /*bottommost_level*/);
+}
+
+// In bottommost level, single deletions earlier than earliest snapshot can be
```
Comment (maysamyabandeh) [inline db/compaction_iterator_test.cc]: While this is correct, the minimum requirement is to keep single deletion larger than than the earliest write-write conflict snapshot. So if there is no write-write conflict snapshot, the SingleDelete could be compacted out anyway.

File: db/compaction_iterator_test.cc [hunk 4]
```diff
@@ -565,6 +649,300 @@ TEST_F(CompactionIteratorTest, SingleMergeOperand) {
...
+    : public CompactionIteratorTest {
+ public:
+  bool UseSnapshotChecker() const override { return true; }
+};
+
+// Uncommitted keys (keys with seq > last_committed_seq) should be output as-is
+// while committed version of these keys should get compacted as usual.
+
+TEST_F(CompactionIteratorWithSnapshotCheckerTest,
+       PreserveUncommittedKyes_Value) {
```
Comment (maysamyabandeh) [inline db/compaction_iterator_test.cc]: s/Kyes_Value/Keys_Value

File: db/compaction_iterator_test.cc [hunk 5]
```diff
@@ -565,6 +649,300 @@ TEST_F(CompactionIteratorTest, SingleMergeOperand) {
...
+  RunTest(
+      {test::KeyStr("foo", 3, kTypeValue), test::KeyStr("foo", 2, kTypeValue),
+       test::KeyStr("foo", 1, kTypeValue)},
```
Comment (maysamyabandeh) [inline db/compaction_iterator_test.cc]: I do not see any snapshot set here. So why is that 1 can go but not 2 or 3?

File: db/compaction_iterator_test.cc [hunk 6]
```diff
@@ -565,6 +649,300 @@ TEST_F(CompactionIteratorTest, SingleMergeOperand) {
...
+           test::KeyStr("foo", 1, kTypeBlobIndex)},
+          {"v4", "v3", "v2", "v1"},
+          {test::KeyStr("foo", 4, kTypeBlobIndex),
+           test::KeyStr("foo", 3, kTypeBlobIndex),
+           test::KeyStr("foo", 1, kTypeBlobIndex)},
+          {"v4", "v3", "v1"}, 3);
+}
+
+// At bottom level, sequence numbers can be zero out, and deletions can be
+// removed, but only when they are visible to earliest sequence.
```
Comment (maysamyabandeh) [inline db/compaction_iterator_test.cc]: s/earliest sequence/earliest snapshot

File: db/compaction_iterator_test.cc [hunk 7]
```diff
@@ -565,6 +649,300 @@ TEST_F(CompactionIteratorTest, SingleMergeOperand) {
...
+           test::KeyStr("a", 1, kTypeValue)},
+          {"", "v1"},
+          {test::KeyStr("a", 2, kTypeSingleDeletion),
+           test::KeyStr("a", 1, kTypeValue)},
+          {"", "v1"}, 2 /*last_committed_seq*/);
+}
+
+// Single delete should be kept in case it is not visible to the
+// earliest write conflict snapshot. In this case, corresponding value
+// can be trim to save space.
```
Comment (maysamyabandeh) [inline db/compaction_iterator_test.cc]: s/In this case/Otherwise
s/trim/trimmed

File: db/compaction_iterator_test.cc [hunk 8]
```diff
@@ -565,6 +649,300 @@ TEST_F(CompactionIteratorTest, SingleMergeOperand) {
...
+TEST_F(CompactionIteratorWithSnapshotCheckerTest, CompactionFilter_Value) {
+  std::unique_ptr<CompactionFilter> compaction_filter(
+      new FilterAllKeysCompactionFilter());
+  RunTest(
+      {test::KeyStr("a", 2, kTypeValue), test::KeyStr("a", 1, kTypeValue),
+       test::KeyStr("b", 3, kTypeValue), test::KeyStr("c", 1, kTypeValue)},
+      {"v2", "v1", "v3", "v4"},
+      {test::KeyStr("a", 2, kTypeValue), test::KeyStr("a", 1, kTypeDeletion),
+       test::KeyStr("b", 3, kTypeValue), test::KeyStr("c", 1, kTypeDeletion)},
+      {"v2", "", "v3", ""}, 1 /*last_committed_seq*/,
```
Comment (maysamyabandeh) [inline db/compaction_iterator_test.cc]: Why the committed keys still exist and yet their values are empty?

File: db/db_iterator_test.cc [hunk 9]
```diff
@@ -18,9 +18,34 @@
 
 namespace rocksdb {
 
-class DBIteratorTest : public DBTestBase {
```
Comment (maysamyabandeh) [inline db/db_iterator_test.cc]: I reintroduced a bug that we fixed before and the db_iterator_test seems passing. Would be great if we add a test that would catch the bug if reoccurs? or even better if they are existing tests that could be extended to catch the bug.

```
--- a/utilities/transactions/write_prepared_txn.cc
+++ b/utilities/transactions/write_prepared_txn.cc
@@ -59,7 +59,7 @@ Iterator* WritePreparedTxn::GetIterator(const ReadOptions& options,
   Iterator* db_iter = wpt_db_->NewIterator(options, column_family);
   assert(db_iter);

-  return write_batch_.NewIteratorWithBase(column_family, db_iter);
+  return write_batch_.NewIteratorWithBase(db_iter);
```

File: db/compaction_iterator_test.cc [hunk 10]
```diff
@@ -159,61 +173,129 @@ class FakeCompaction : public CompactionIterator::CompactionProxy {
...
+      bool bottommost_level = false,
+      SequenceNumber earliest_write_conflict_snapshot = kMaxSequenceNumber) {
     std::unique_ptr<InternalIterator> range_del_iter(
         new test::VectorIterator(range_del_ks, range_del_vs));
     range_del_agg_.reset(new RangeDelAggregator(icmp_, snapshots_));
     ASSERT_OK(range_del_agg_->AddTombstones(std::move(range_del_iter)));
 
     std::unique_ptr<CompactionIterator::CompactionProxy> compaction;
-    if (filter) {
+    if (filter || bottommost_level) {
```
[the remark this answers is not in the document]
Comment (yiwu-arbug) [reply db/compaction_iterator_test.cc]: To have a fake compaction object to tell CompactionIterator it is at bottom level.

File: db/compaction_iterator_test.cc (hunk 1 above)
Comment (yiwu-arbug) [reply db/compaction_iterator_test.cc]: Additional message when the assert fail. Useful to locate the failure when the test fails.

File: db/compaction_iterator_test.cc (hunk 3 above)
Comment (yiwu-arbug) [reply db/compaction_iterator_test.cc]: Right.. I leave a comment in db/compaction_iterator.cc in this PR about it. Maybe we can address it when someone want to refactor compaction filter later on.

File: db/compaction_iterator_test.cc (hunk 5 above)
Comment (yiwu-arbug) [reply db/compaction_iterator_test.cc]: The last_committed_seq is set to 2 (the last param of RunTest() call). I'll add comment.

File: db/compaction_iterator_test.cc (hunk 8 above)
Comment (yiwu-arbug) [reply db/compaction_iterator_test.cc]: Deletions don't come with values.

File: db/db_iterator_test.cc (hunk 9 above)
Comment (yiwu-arbug) [reply db/db_iterator_test.cc]: This bug is on a higher layer. We need to add the test to write_prepared_transaction_test.cc.

**Pull request closed without merging**